### ISSUE TITLE
Many to many pickup-delivery problem question

### ISSUE THREAD

**Issue by habimt** [VROOM-Project/vroom#366]
Hello

I have the following problem:
* There is one vehicle which starts empty and there are many pickup points which contain one or many items that have to be delivered to delivery points while those delivery points require items from one or many pickup points.
* Every pickup point contains unique items while multiple delivery points can request same items.
* Vehicle capacity is currently not an issue.

To Illustrate my point, let's assume we have following pickup points (P) containing items (alphabet letters) and delivery points (D) which require those items:

P1: a, b
P2: c
P3: d
D1: a
D2: b, d
D3: a, c

As we can see, in order for vehicle to be eligible to deliver to point D1 it has to stop at P1 where it will pickup items a (2x) and b (1x) (it is assumed that pickup points have the required quantity of multiple same items). At this point, vehicle could deliver either to D1 or make another pickup (for example at P2 and then become eligible to deliver to D3).

Can VROOM take this into account and decide which routing choice is more optimal?.

So far I have tried the following:
* Create multiple shipments for every delivery-pickup relation. This works only if I have one delivery location, similar as in #313 in which case I multiple deliveries with same location. This however does not work if I add multiple delivery locations (I get HTTP 200 with empty response).
* Create multiple jobs with delivery and pickup arrays, this didn't appear to work and as It clearly status in documentation: '_It is assumed that all delivery-related amounts for jobs are loaded at vehicle start, while all pickup-related amounts for jobs are brought back at vehicle end._'. This does not seem to be solution to his problem since my vehicle starts empty, makes pickups and deliveries and is then empty again at the end.
* Create multiple jobs with higher priority for all pickup related jobs. This doesn't solve my problem since it forces the vehicle to make all pickups first despite the fact that it may become eligible for some deliveries while it completes some pickup jobs.

Is VROOM able to solve this routing issue for me? If not I would gladly appreciate if someone could point me to the right direction. Thanks in advance and cheers!

**Comment by jcoupey** [comment]
In any case, as stated in https://github.com/VROOM-Project/vroom/issues/313#issuecomment-599425865, the right way to model doing pickup and deliveries along a route is by defining `shipment` objects.

Creating shipments for every delivery-pickup relation you have is probably what you should do, but it does not exactly match your reasoning of having a single vehicle picking up everything required before a delivery. Is there a hard constraint related to your use-case that requires this?

In other words: considering the needs at D3, you seem to expect that a single vehicle will go to both P1 and P2 (in whatever order) then deliver to D3. But it may be acceptable (and even cheaper) to have two different vehicles go P1->D3 and P2->D3.

**Comment by habimt** [comment]
> But it may be acceptable (and even cheaper) to have two different vehicles go P1->D3 and P2->D3.

This is unfortunately unacceptable in my case. Every delivery is actually an order created by customer which has receipt containing all items which customer has ordered so they have to be delivered together in single delivery.

There is a possibility that there will be multiple vehicles but in that case one vehicle would still have to deliver all required items to single customer (though perhaps they could meet at certain point and exchange pickup items to increase efficiency). Nevertheless, for now there is only one vehicle so I have to solve the problem with one vehicle only.

As I stated above, I get response with empty content (I'm using openrouteservice) when I create shipments for every delivery-pickup relation as below:

```json
{
  "vehicles": [
    {
      "id": 1,
      "start": [
        8.751001121009775,
        49.45888317507314
      ],
      "end": [
        8.751001121009775,
        49.45890298342954
      ],
      "profile": "driving-car"
    }
  ],
  "shipments": [
    {
      "pickup": {
        "id": 100,
        "location": [
          8.74955901181615,
          49.43044850911426
        ]
      },
      "delivery": {
        "id": 0,
        "location": [
          8.751753190378604,
          49.4399611152364
        ]
      }
    },
    {
      "pickup": {
        "id": 100,
        "location": [
          8.74955901181615,
          49.43044850911426
        ]
      },
      "delivery": {
        "id": 1,
        "location": [
          8.667643012151231,
          49.45058134510214
        ]
      }
    },
    {
      "pickup": {
        "id": 300,
        "location": [
          8.738344321385835,
          49.46611126961545
        ]
      },
      "delivery": {
        "id": 1,
        "location": [
          8.667643012151231,
          49.45058134510214
        ]
      }
    },
    {
      "pickup": {
        "id": 100,
        "location": [
          8.74955901181615,
          49.43044850911426
        ]
      },
      "delivery": {
        "id": 2,
        "location": [
          8.722741273830609,
          49.470072197224226
        ]
      }
    },
    {
      "pickup": {
        "id": 200,
        "location": [
          8.681783273998152,
          49.42727723041344
        ]
      },
      "delivery": {
        "id": 2,
        "location": [
          8.722741273830609,
          49.470072197224226
        ]
      }
    }
  ]
}
```
I will however get a proper response in case like the following:
```json
{
  "vehicles": [
    {
      "id": 1,
      "start": [
        8.710988378667798,
        49.71586296874179
      ],
      "end": [
        8.711261201781692,
        49.716921366732834
      ],
      "profile": "driving-car"
    }
  ],
  "shipments": [
    {
      "pickup": {
        "id": 100,
        "location": [
          8.692436406923006,
          49.71215839411545
        ]
      },
      "delivery": {
        "id": 0,
        "location": [
          8.622048043538362,
          49.6945136857467
        ]
      }
    },
    {
      "pickup": {
        "id": 100,
        "location": [
          8.692436406923006,
          49.71215839411545
        ]
      },
      "delivery": {
        "id": 1,
        "location": [
          8.617955696829952,
          49.70827710808521
        ]
      }
    },
    {
      "pickup": {
        "id": 300,
        "location": [
          8.647693416244396,
          49.693631282112136
        ]
      },
      "delivery": {
        "id": 1,
        "location": [
          8.617955696829952,
          49.70827710808521
        ]
      }
    },
    {
      "pickup": {
        "id": 100,
        "location": [
          8.692436406923006,
          49.71215839411545
        ]
      },
      "delivery": {
        "id": 2,
        "location": [
          8.61304488077986,
          49.72203663183913
        ]
      }
    },
    {
      "pickup": {
        "id": 200,
        "location": [
          8.668155149786443,
          49.70086651897759
        ]
      },
      "delivery": {
        "id": 2,
        "location": [
          8.61304488077986,
          49.72203663183913
        ]
      }
    }
  ]
}
```

The difference between the two mentioned requests is that in the first example, I placed pickups near deliveries while in the second example I intentionally put all pickups on the same road which leads to deliveries so in that scenario it would be optimal for vehicle to take all pickups first and then deliver them. Second example is not what I'm interested in because in my scenario there is a high probability that some deliveries will be close to pickups, so the vehicle could become eligible for delivery if it visited certain pickup spots and then make that delivery before other pickups if that is more optimal route.

Am I doing something wrong? Any suggestions would be appreciated.

**Comment by jcoupey** [comment]
If you actually need **a chain** of precedence constraints (several pickups to be done before a single delivery), then no you can't model that with our current codebase.

On the other hand, when using multiple pickup-delivery pairs with same-place pickups/deliveries, they will tend to be grouped in a row, simply because it's cheaper that way. Plus a target delivery that have many "parent" pickups will tend to be done only when those are already all visited, because else it would mean delivering several times at the same place, so would be more expensive.

As a matter of fact, when I solve your first example against the demo server, I'm getting exactly this behaviour (not sure what you mean by "response with empty content").

```bash
$ curl --header "Content-Type:application/json" --data @input_1.json http://solver.vroom-project.org | jq '.routes[0].steps[] | [.type,.id] | @csv'
"\"start\","
"\"pickup\",100"
"\"pickup\",100"
"\"pickup\",100"
"\"delivery\",0"
"\"pickup\",300"
"\"delivery\",1"
"\"delivery\",1"
"\"pickup\",200"
"\"delivery\",2"
"\"delivery\",2"
"\"end\","
```

**Comment by habimt** [comment]
> As a matter of fact, when I solve your first example against the demo server, I'm getting exactly this behaviour (not sure what you mean by "response with empty content").

What I mean is, when I query `https://api.openrouteservice.org/optimization` with the first JSON example, I get empty response, no JSON data or anything. I do get proper JSON response when I run second example I posted. Now I also tried first example on the test server you mentioned and got a valid response so perhaps openrouteservice is running an older version of VROOM where my first example does not appear to work.

**Comment by nilsnolde** [comment]
Yep, ORS is running vroom 1.6.0. Drop a comment on https://ask.openrouteservice.org to get their attention.

**Comment by jcoupey** [comment]
I don't see any reason for a difference of behaviour between the two examples, even with `v1.6.0`.

Also an empty response should not happen. @habimt maybe you could try with a local ORS stack for more debug feedback?

**Comment by habimt** [comment]
I have managed to build and run vroom 1.7+osrm and am now using vroom-express for api calls.

This appears to work as I get proper response in both examples and routing steps seem to make sense.  I have also tried to build vroom v1.6.0 to check if my first json would work there, however I got some sort of OSRM related error when building 1.6.0 (following wiki steps for building 1.7.0) so I can't give you any more feedback related to the empty response issue.

**Comment by jcoupey** [comment]
If you want to build `v1.6.0`, you need to `apt install libboost-all-dev` (this dependency has been dropped for `v1.7.0`). You *should* not see any difference in behaviour between those two versions on top of OSRM for your instance example, but it never hurts to check.

I suspect the empty response mentioned above is related to the ORS + VROOM setup in place on the Openrouteservice API. I'm not sure if we can get some more debug info from there, but it would be worth reporting to try and get to the bottom of the problem. @nilsnolde what would be the best place to report the empty response problem?

**Comment by nilsnolde** [comment]
Empty response is very strange.. I just tried myself and had then a look into the server logs. It's causing a dreaded `segmentation fault`:
```
Mon, 20 Jul 2020 15:15:21 GMT: Segmentation fault (core dumped)
```

I'm not working for ORS anymore. I just asked if it's ok if I quickly upgrade to our new vroom docker image regardless, I still got access.

Generally ask stuff which is not core ORS in the forum: https://ask.openrouteservice.org.

**Comment by jcoupey** [comment]
My bad, I forgot about #333 which is most probably what you've been hitting here. Upgrading should definitly solve the mentioned segfaulft on the ORS side.

**Comment by nilsnolde** [comment]
Ok, I upgraded the ORS side of things. Your first example runs now @habimt

**Comment by jcoupey** [comment]
We should probably close now?
